docs/tutorial/algorithm_interfaces.rst

```diff
@@ -1,2 +1,464 @@
 Algorithm interfaces
-====================
\ No newline at end of file
+====================
+
+In the previous chapter, we went over the :ref:`big picture of ALF <chap2_big_picture>`.
+Now we will take a look at how different algorithm interfaces interact with each
+other in the on-policy/off-policy pipelines. We will try modifying some
+algorithm interfaces of the previous example :doc:`./a_minimal_working_example`,
+for an illustration purpose. Then we introduce a simple off-policy training
+example and again explore its algorithm interfaces.
+
+Interface Overview
+------------------
+
+.. image:: images/algorithm_interfaces.png
+    :width: 800
+    :align: center
+    :alt: interaction between algorithm interfaces
+
+
+There are several key interfaces (highlighted in the figure above) that play
+important roles in assembling a training pipeline of an ALF algorithm, and their
+signatures are summarized below.
+
+======================================================  ===========================================  =======================================  =============================== ===============
+Interface                                               Tensor shape                                 Input                                    Output                          Which pipeline?
+======================================================  ===========================================  =======================================  =============================== ===============
+:meth:`~.AlgorithmInterface.predict_step`               :math:`[1,\ldots]`                           ``inputs``, ``state``                    ``output``, ``state``, ``info`` On/Off-policy
```

Comment:
Not related to this tutorial, but I think one potential approach to lower the learning cost of the first time user and also the cost of composing tutorial is to unify the names of the lifecycle hooks, i.e. `xxx_step` or `xxx_hook` or `on_xxx_hook`.

Reply:
I think the reason why we have "step" in some hooks but not in others is that "predict_step", "rollout_step" and "train_step" are indeed called on a per-step basis, i.e., t, t+1, t+2, ... But for "calc_loss", "after_udpate", "after_train_iter" or other hooks they are not called on a per-step basis. So "step" here actually means the time step concept in RL.

Reply:
Got it. That totally make sense - worth explaining in the tutorial as well?

Reply:
Sure, I've explained their difference in temporal logic. But I won't talk about their naming logic in the tutorial.